### PR TITLE
Add support for unknown experimental s2k's

### PR DESCRIPTION
**Pull request by ni4**
Since unknown s2k may include not-yet-supported secret key fields, those should not be discarded during the load/save.
So, this PR:
- adds support of loading/saving of unknown s2k
- adds possible 'Unknown' result for `rnp_key_get_protection_type()`
- refactors code, which related to s2k storage, to C++ style, so we could rely on constructors/destructors.

Fixes partially #1303

### CODE REVIEW

**Review by antonsviridenko — APPROVED** [review]

File: src/lib/rnp.cpp [hunk 1]
```diff
@@ -2477,32 +2441,25 @@ pgp_write_handler(pgp_password_provider_t *pass_provider,
...
-        if (!osig->create_set) {
-            sinfo.sigcreate = ctx->sigcreate;
-        }
-
-        if (!list_append(&ctx->signers, &sinfo, sizeof(sinfo))) {
-            return RNP_ERROR_OUT_OF_MEMORY;
+        if (!sig.create_set) {
+            sinfo.sigcreate = ctx.sigcreate;
         }
+        ctx.signers.emplace_back(sinfo);
```
Comment (rrrooommmaaa) [inline src/lib/rnp.cpp:2461]: why not `push_back`?

File: src/librepgp/stream-ctx.cpp [hunk 2]
```diff
@@ -74,19 +75,16 @@ rnp_ctx_add_encryption_password(rnp_ctx_t *    ctx,
     if (!pgp_s2k_derive_key(&info.s2k, password, info.key, sizeof(info.key))) {
         return RNP_ERROR_GENERIC;
     }
-    if (!list_append(&ctx->passwords, &info, sizeof(info))) {
-        pgp_forget(&info, sizeof(info));
+    try {
+        ctx.passwords.emplace_back(info);
```
Comment (rrrooommmaaa) [inline src/librepgp/stream-ctx.cpp:79]: why not `push_back`?

File: src/librepgp/stream-write.cpp [hunk 3]
```diff
@@ -1644,8 +1633,8 @@ init_literal_dst(pgp_write_handler_t *handler, pgp_dest_t *dst, pgp_dest_t *writ
 {
     pgp_dest_packet_param_t *param;
     rnp_result_t             ret = RNP_ERROR_GENERIC;
-    int                      flen = 0;
     uint8_t                  buf[4];
+    size_t                   flen = 0;
```
Comment (rrrooommmaaa) [inline src/librepgp/stream-write.cpp:1637]: why is this moved?

File: src/librepgp/stream-ctx.cpp (hunk 2 above)
Comment (ni4) [reply src/librepgp/stream-ctx.cpp:79]: @rrrooommmaaa In my understanding `emplace_back()` would directly construct object here, using the move constructor, while `push_back()` will first create temporary and then do copy/move. Is it wrong in this  case?

File: src/librepgp/stream-write.cpp (hunk 3 above)
Comment (ni4) [reply src/librepgp/stream-write.cpp:1637]: @rrrooommmaaa Just typo - first attempted to move it below, where it is used for the first time, but then realized that there is a goto before. Should I revert this to the initial location?

Comment (rrrooommmaaa) [reply src/librepgp/stream-write.cpp:1637]: @ni4 would be much easier to track changes, evidently

File: src/librepgp/stream-ctx.cpp (hunk 2 above)
Comment (rrrooommmaaa) [reply src/librepgp/stream-ctx.cpp:79]: @ni4 Just checking. We have lots of `push_back` calls throughout the app.
We even have `siginfos.push_back({})` which should probably be replaced with `emplace_back`
I'm not an expert in this field, but from what I read, `emplace_back` is useful when you pass a list of constructor arguments to it, so that it is constructed in place. But since `sinfo` is the object itself, I suspect "move construction" will happen in the same way as in `push_back` implementation. I'm just trying to figure out the general rule to use throughout the app.

Comment (ni4) [reply src/librepgp/stream-ctx.cpp:79]: @rrrooommmaaa My opinion on this is like 'call emplace_back() when we move object/construct from the parameter list or push_back() when we copy', but it could be subject to change if there are some arguments I'm missing. Maybe @dewyatt has also some thoughts on this topic?

File: src/librepgp/stream-write.cpp (hunk 3 above)
Comment (ni4) [reply src/librepgp/stream-write.cpp:1637]: Okk, will re-push.

File: src/librepgp/stream-ctx.cpp (hunk 2 above)
Comment (dewyatt) [reply src/librepgp/stream-ctx.cpp:79]: I don't think there's an advantage to `emplace_back` in this case since we're inserting an existing object, not constructing and inserting. Since `emplace*` can have surprising behavior with implicit conversions, it may be best to stick with `push_back` except in cases where there is a clear advantage. (Effective Modern C++ has some guidance in Item 42).

Comment (ni4) [reply src/librepgp/stream-ctx.cpp:79]: @dewyatt Thanks, let's stick to `push_back()` in such cases then.

Comment (ni4) [reply src/librepgp/stream-ctx.cpp]: @rrrooommmaaa Changed to `push_back()` and re-pushed. Any other comments?

**Review by rrrooommmaaa — APPROVED** [review]
LGTM!
